### PR TITLE
feat: edit CIDR for RstudioV2

### PR DESCRIPTION
**Pull request by manikandan-thangavelu-rl**
Issue #, if available:

Description of changes:

Modified the edit CIDR feature to be in compatible with RstudioV2 

Checklist:

- [x] Have you successfully deployed to an AWS account with your changes?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully tested with your changes locally?

AS review ticket id:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js [hunk 1]
```diff
@@ -338,6 +337,80 @@ class ALBService extends Service {
...
+  async modifyRule(requestContext, resolvedVars) {
+    const subdomain = this.getHostname(resolvedVars.prefix, resolvedVars.envId);
+    try {
+      const cidrLen = resolvedVars.cidr.length;
+      // ModifyRule does not accept the empty value to update
+      // so the system should validate the cidr and if it is empty then
+      // replace the default ip "0.0.0.0/0"
+      if (cidrLen === 0) {
+        resolvedVars.cidr = ['0.0.0.0/0'];
+      }
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js]: Defaulting to `0.0.0.0` would be a security concern. Can we throw an exception instead?

File: addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js [hunk 2]
```diff
@@ -338,6 +337,80 @@ class ALBService extends Service {
...
+
+  /**
+   * Method to modify rule. The method modify the rule using the ALB SDK client.
+   * Tags are read form the resolvedVars so the billing will happen properly
+   *
+   * @param requestContext
+   * @param resolvedVars
+   * @returns {Promise<D & {$response: Response<D, E>}>}
+   */
+  async modifyRule(requestContext, resolvedVars) {
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js:377]: Can concurrent requests by say the workspace owner (researcher) and an admin cause one of the requests to be omitted. If yes, can we add some locking mechanism to ensure that concurrent requests don't overwrite the rules.

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js [hunk 3]
```diff
@@ -113,26 +122,62 @@ class EnvironmentScCidrService extends Service {
...
+      cloneUpdateRequest = JSON.parse(cloneUpdateRequest);
+      existingEnvironment.cidr = cloneUpdateRequest;
+    }
     // Write audit event
     await this.audit(requestContext, { action: 'update-environment-sc-cidr', body: existingEnvironment });
-
     return existingEnvironment;
   }
 
+  async modifyELBRule(existingEnvironment, updateRequest, albService, requestContext) {
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js:149]: Same comment that I made about modifyRule function

File: addons/addon-base-raas/packages/base-raas-services/lib/alb/__tests__/alb-service.test.js [hunk 4]
```diff
@@ -430,4 +430,63 @@ describe('ALBService', () => {
       }
     });
   });
+
+  describe('modifyRule', () => {
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/alb/__tests__/alb-service.test.js:465]: We should verify if we are calling albClient.modifyRule with correct parameters

File: addons/addon-base-raas/packages/base-raas-services/lib/alb/__tests__/alb-service.test.js [hunk 5]
```diff
@@ -430,4 +430,63 @@ describe('ALBService', () => {
...
+          },
+        };
+      });
+      service.getAlbSdk = jest.fn().mockResolvedValue(albClient);
+      await service.modifyRule({}, { cidr: [], projectId: '' });
+      expect(albClient.modifyRule).toHaveBeenCalled();
+    });
+  });
+
+  describe('describeRules', () => {
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/alb/__tests__/alb-service.test.js:536]: Can we test non-empty response use case as well

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js (hunk 3 above)
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js:149]: Can we add a unit test for this method?

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js [hunk 6]
```diff
@@ -343,8 +344,23 @@ class EnvironmentScService extends Service {
...
+        const albService = await this.service('albService');
+        const resolvedVars = { ruleARN: ListenerRuleARN, projectId: env.projectId };
+        const ruleSourceIps = await albService.describeRules(requestContext, resolvedVars);
+        currentIngressRules.map(obj => {
+          if (obj.fromPort === 443) {
+            obj.cidrBlocks = ruleSourceIps;
+          }
+          return obj;
+        });
+      }
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js]: Can we add a unit test for this branch?

Comment (narendran1560) [reply addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-service.js]: Separate method has been created for the describeELBRule and incorporated the test cases as well. Please review

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js (hunk 3 above)
Comment (narendran1560) [reply addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js:149]: Created the new test function for the modifyELBRule. Please review

File: addons/addon-base-raas/packages/base-raas-services/lib/alb/__tests__/alb-service.test.js (hunk 5 above)
Comment (narendran1560) [reply addons/addon-base-raas/packages/base-raas-services/lib/alb/__tests__/alb-service.test.js:536]: Yes. Added non-empty response use case as well. Please review

File: addons/addon-base-raas/packages/base-raas-services/lib/alb/__tests__/alb-service.test.js (hunk 4 above)
Comment (narendran1560) [reply addons/addon-base-raas/packages/base-raas-services/lib/alb/__tests__/alb-service.test.js:465]: Added and verified the params of the albClient.modifyRule. Please review

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js (hunk 3 above)
Comment (narendran1560) [reply addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js:149]: modifyELBRule function declaration as included into the existing locking mechanism. Please review

File: addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js (hunk 2 above)
Comment (narendran1560) [reply addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js:377]: modifyRule function declaration as included into the existing locking mechanism. Please review

File: addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js (hunk 1 above)
Comment (narendran1560) [reply addons/addon-base-raas/packages/base-raas-services/lib/alb/alb-service.js]: Yes. Removed the default ip address 0.0.0.0 and added the appropriate error message in the exception.

**Review by jn1119 — APPROVED** [review]

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js [hunk 7]
```diff
@@ -113,26 +122,64 @@ class EnvironmentScCidrService extends Service {
...
+        };
+        await albService.modifyRule(requestContext, resolvedVars);
+        // Removed the original value in 443 Port CIDRs and replace the default values in the updateRequest
+        // because the new RStudio ALB we are not storing 443 Port CIDRs into security group.
+        updateRequest.map(obj => {
+          if (obj.fromPort === 443) {
+            obj.cidrBlocks = ['0.0.0.0/0'];
+          }
+          return obj;
+        });
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js:177]: Didn't quite catch this part. Where do we use this updateRequest? It doesn't seem like we return it. And why `0.0.0.0`?

**Pull request merged**